src/parser/mod.rs

```diff
@@ -15017,6 +15026,48 @@ impl<'a> Parser<'a> {
         }
     }
 
+    fn parse_go(&mut self) -> Result<Statement, ParserError> {
+        // previous token should be a newline (skipping non-newline whitespace)
```

Comment:
hmm I didn't look too closely but not sure I followed the function body on a high level, from the docs and the representation in the parser the Go statement only contains an optional int, so that I imagined all the function needs to do would be to `maybe_parse(|parser| parser.parse_number_value())` or similar? The function is currently managing whitespace and semicolon but not super clear to me why that is required

Reply:
Does [my comment above](https://github.com/apache/datafusion-sqlparser-rs/pull/1809#discussion_r2042359100) & the code comment below, "using this peek function because we want to halt this statement parsing upon newline" help?

```mssql
-- this should parse as `Go { Some(4) }`
GO 4`

-- this should parse as something like `Go { None }`, `Expr(4)`
GO
4
```

Eg, the whitespace is meaningful for parsing

Reply:
Ah interesting, I think it would be very helpful to add this example to the code to illustrate the implications,

Then for the function, I think using the `peek_token_no_skip` could simplify the impl. Something roughly like the following? to avoid manually tracking token counts and lookbacks
```rust
let count = loop {
  match self.peek_token_no_skip() {
    Token::Whitespace(Whitespace::Newline) | Token::Whitespace(Whitespace::SingleLineComment) | Token::EOF => {
      self.advance_token();
      break None
    }
    Token::Whitespace(_) => {
      self.advance_token();
    }
    _ => {
      break Some(self.parse_number())
    }
  }
};
```

Reply:
Yes, I have something like that already for parsing the batch count, which doesn't need to look backwards and already uses peek_token_no_skip. I think between here & your other comment on the alias function, the concern is regarding the newline logic which looks backwards. I'll try and move that into a helper so it can be simplified.